### PR TITLE
feat(image): updated json report added package locations fields

### LINKED ISSUE #4418: Include line numbers in json output
## Description
Line numbers are included in SARIF, but JSON. It would be nice if they are included in JSON as well.

### PR DESCRIPTION
**Pull request by parvez0**
## Description
This pull request introduces a new feature to include package location indexes in the JSON report.

**Changes**
1. Modified `pkg/scanner/local/server.go` to use the location data from the detected vulnerability.
2. Updated `pkg/report/sarif.go` to reuse the calculated location information if available.
3. Added a new key `Location` under `DetectedVulnerability` in `pkg/types/vulnerability.go` to update the extracted location data.
4. Updated `integration/testdata/*.json.golden` to verify Location field in json reports

**Usage**
```sh
$ git clone https://github.com/knqyf263/trivy-ci-test.git
$ cd trivy-ci-test
$ trivy fs --format template --format json -o trivy_report.json --scanners vuln .
```

**Command output before this feature change**
```json
{
  "SchemaVersion": 2,
  "CreatedAt": "2024-03-16T21:54:44.700427+05:30",
  "ArtifactName": ".",
  "ArtifactType": "filesystem",
  "Metadata": {
    ...
  },
  "Results": [
    {
      "Target": "Cargo.lock",
      "Class": "lang-pkgs",
      "Type": "cargo",
      "Packages": [...],
      "Vulnerabilities": [
        {
          "VulnerabilityID": "CVE-2019-15542",
          "PkgID": "ammonia@1.9.0",
          "PkgName": "ammonia",
          "PkgIdentifier": {
            "PURL": "pkg:cargo/ammonia@1.9.0"
          },
          "InstalledVersion": "1.9.0",
          "FixedVersion": "2.1.0",
          "Status": "fixed",
          "Layer": {},
          "SeveritySource": "ghsa",
          "PrimaryURL": "https://avd.aquasec.com/nvd/cve-2019-15542",
          "DataSource": {...},
          "Title": "Uncontrolled recursion in ammonia",
          "Description": "An issue was discovered in the ammonia crate before 2.1.0 for Rust. There is uncontrolled recursion during HTML DOM tree serialization.",
          "Severity": "HIGH",
          "CweIDs": ["CWE-674],
          "VendorSeverity": {...},
          "CVSS": {...},
          "References": [...],
          "PublishedDate": "2019-08-26T18:15:12.467Z",
          "LastModifiedDate": "2020-08-24T17:37:01.14Z"
        }
      ]
    }
  ]
}
```
**Command output after introducing this feature**
```json
{
  "SchemaVersion": 2,
  "CreatedAt": "2024-03-16T21:54:44.700427+05:30",
  "ArtifactName": ".",
  "ArtifactType": "filesystem",
  "Metadata": {
    ...
  },
  "Results": [
    {
      "Target": "Cargo.lock",
      "Class": "lang-pkgs",
      "Type": "cargo",
      "Packages": [...],
      "Vulnerabilities": [
        {
          "VulnerabilityID": "CVE-2019-15542",
          "PkgID": "ammonia@1.9.0",
          "PkgName": "ammonia",
          "PkgIdentifier": {
            "PURL": "pkg:cargo/ammonia@1.9.0"
          },
          "InstalledVersion": "1.9.0",
          "FixedVersion": "2.1.0",
          "Status": "fixed",
          "Layer": {},
          "SeveritySource": "ghsa",
          "PrimaryURL": "https://avd.aquasec.com/nvd/cve-2019-15542",
          "Locations": [
            {
              "StartLine": 2,
              "EndLine": 13
            }
          ],
          "DataSource": {...},
          "Title": "Uncontrolled recursion in ammonia",
          "Description": "An issue was discovered in the ammonia crate before 2.1.0 for Rust. There is uncontrolled recursion during HTML DOM tree serialization.",
          "Severity": "HIGH",
          "CweIDs": [ "CWE-674"],
          "VendorSeverity": {...},
          "CVSS": {...},
          "References": [...],
          "PublishedDate": "2019-08-26T18:15:12.467Z",
          "LastModifiedDate": "2020-08-24T17:37:01.14Z"
        }
      ]
    }
  ]
}

```

## Related issues
- Close #4418 

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/scanner/local/scan.go [hunk 1]
```diff
@@ -158,6 +158,9 @@ func (s Scanner) ScanTarget(ctx context.Context, target types.ScanTarget, option
 		s.vulnClient.FillInfo(results[i].Vulnerabilities)
 	}
 
+	// Fill package location information
+	fillPackageLocation(results)
```
Comment (DmitriyLewen) [inline pkg/scanner/local/scan.go:162]: `Package` already contains location - https://github.com/aquasecurity/trivy/blob/317792433e4e961441f772c6bd22d63873a8c986/pkg/fanal/types/artifact.go#L101

what if we just add location when creating a new vulnerability?
https://github.com/aquasecurity/trivy/blob/317792433e4e961441f772c6bd22d63873a8c986/pkg/detector/library/detect.go#L33-L37

File: pkg/report/sarif_test.go [hunk 2]
```diff
@@ -541,6 +541,176 @@ func TestReportWriter_Sarif(t *testing.T) {
...
+						"debian:9",
+					},
+					RepoDigests: []string{
+						"debian@sha256:a8cc1744bbdd5266678e3e8b3e6387e45c053218438897e86876f2eb104e5534",
+					},
+				},
+				Results: types.Results{
+					{
+						Target: "library/test",
+						Class:  types.ClassOSPkg,
```
Comment (DmitriyLewen) [inline pkg/report/sarif_test.go:560]: IIRC OS package don't have locations

File: pkg/report/sarif_test.go [hunk 3]
```diff
@@ -541,6 +541,176 @@ func TestReportWriter_Sarif(t *testing.T) {
...
+								},
+								Locations: []ftypes.Location{
+									{
+										StartLine: 15,
+										EndLine:   20,
+									},
+									{
+										StartLine: 30,
+										EndLine:   35,
+									},
```
Comment (DmitriyLewen) [inline pkg/report/sarif_test.go:612]: location from package is different- https://github.com/aquasecurity/trivy/blob/80bdd63cf850e7abc8e1101f12cfced798a4dd78/pkg/report/sarif_test.go#L46-L55

**Review by DmitriyLewen — COMMENTED** [review]
Hello @parvez0 
Thanks for your work!

Left comments.

Regards, Dmitriy

File: pkg/scanner/local/scan.go (hunk 1 above)
Comment (parvez0) [reply pkg/scanner/local/scan.go:162]: > Package already contains location -
> 
> [trivy/pkg/fanal/types/artifact.go](https://github.com/aquasecurity/trivy/blob/317792433e4e961441f772c6bd22d63873a8c986/pkg/fanal/types/artifact.go#L101)
> 
> Line 101 in [3177924](https://github.com/aquasecurity/trivy/commit/317792433e4e961441f772c6bd22d63873a8c986)
> 
>  Locations []Location `json:",omitempty"`
Not clear on this should I create a new type for Locations?

File: pkg/report/sarif_test.go (hunk 2 above)
Comment (parvez0) [reply pkg/report/sarif_test.go:560]: Not sure, I copied the first test case and used it, it has location fields in packages but if you want I change it as well. Also it will be great if you can point out any relevant reading materials that I can go through.

Comment (DmitriyLewen) [reply pkg/report/sarif_test.go:560]: You can see the packet parser [here](https://github.com/aquasecurity/trivy/tree/main/pkg/fanal/analyzer/pkg).

You can see the language files that support places [here](https://aquasecurity.github.io/trivy/v0.50/docs/coverage/language/)

Comment (DmitriyLewen) [reply pkg/report/sarif_test.go:560]: > I copied the first test case and used it,

I think it's better to add language package (like yarn.lock or pom.xml) to avoid confusion when we support this code later.

File: pkg/scanner/local/scan.go (hunk 1 above)
Comment (DmitriyLewen) [reply pkg/scanner/local/scan.go]: No. You can use `ftypes.Location` in `VulnerabilityDetected`.

You have already chosen right solution:
https://github.com/aquasecurity/trivy/blob/d1945dd4e3fe8355621ed2aea65d411efd7fc71e/pkg/types/vulnerability.go#L22

File: pkg/report/sarif.go [hunk 4]
```diff
@@ -341,10 +341,20 @@ func ToPathUri(input string, resultClass types.ResultClass) string {
 	return strings.ReplaceAll(strings.ReplaceAll(input, "\\", "/"), "git::https:/", "")
 }
 
-func (sw *SarifWriter) getLocations(name, version, path string, pkgs []ftypes.Package) []location {
+func (sw *SarifWriter) getLocations(name, version, path string, pkgs []ftypes.Package, vulnLocations []ftypes.Location) []location {
```
Comment (DmitriyLewen) [inline pkg/report/sarif.go:344]: We don't need location cache and package check now.

What do you thin about this update for this function:

```go
func getLocations(vulnLocations []ftypes.Location) []location {
	if len(vulnLocations) == 0 {
		return nil
	}

	var locs []location
	for _, loc := range vulnLocations {
		locs = append(locs, location{
			startLine: loc.StartLine,
			endLine:   loc.EndLine,
		})
	}
	return locs
}
```

File: pkg/report/sarif_test.go (hunk 3 above)
Comment (DmitriyLewen) [reply pkg/report/sarif_test.go:612]: You need to add locations in both places:
- Package
- DetectVulnerability
But these must be the same line numbers.

File: pkg/report/sarif.go (hunk 4 above)
Comment (parvez0) [reply pkg/report/sarif.go:344]: @DmitriyLewen Initially, I considered this approach. However, after reviewing the [test cases](https://github.com/aquasecurity/trivy/blob/main/pkg/report/sarif_test.go#L25-L184), which uses `dbTypes.Vulnerabilities` instead of `types.DetectedVulnerabilities`, I decided that implementing backward compatible changes would be beneficial. If you're confident that this won't impact the approach, I'll proceed with updating it to match the example you provided.

Comment (DmitriyLewen) [reply pkg/report/sarif.go:344]: If another app generates `Report` using Trivy packages - these changes will have no effect.
But if this app creates `Report` on its own, it will need to simply copy `Location` from `Package` to `Vulnerabilities Detected`.

So I think these changes won't be a problem.

**Comment by parvez0** [comment]
Hey @DmitriyLewen, needed your opinion on this. After integrating the Location Changes, I noticed that `integration_test/client_server.go` started failing for the `busybox_with_cargo.lock` and `scan pox.xml` test cases. The issue from these test cases is expecting `DetectedVulnerability.Location` to be nil. However, the output golden files `busybox-with-lockfile.json.golden` and `pom.json.golden` are also used in other integration tests where non-nil Location data is expected. To address this, I created two duplicate files that includes Location data. Please let me know if you have a better solution for this.

File: pkg/report/sarif_test.go (hunk 3 above)
Comment (parvez0) [reply pkg/report/sarif_test.go]: It appears there might be referring to wrong test case in the comment. My changes actually begin from line 544. Nevertheless, I've refactored the test case for a package vulnerability as per your suggestion.

File: pkg/report/sarif_test.go (hunk 2 above)
Comment (parvez0) [reply pkg/report/sarif_test.go]: I have added a YARN vulnerability as a test case replacing the previous OSDependency

**Comment by DmitriyLewen** [comment]
@parvez0 You need to update `rpc` package.
Let me know when you're done making changes (to avoid conflicts) and i will update client/server part.

**Comment by parvez0** [comment]
> @parvez0 You need to update `rpc` package. Let me know when you're done making changes (to avoid conflicts) and i will update client/server part.

Hey @DmitriyLewen what do I have to do in `rpc` package I haven't seen any changes that need to be made there, can you please let me know I'll try to do it as soon as possible

**Comment by DmitriyLewen** [comment]
Hm... we don't insert `Locations` in client/server mode...
I created #6366 for this bug.
After merge #6366 you need to merge `main` branch into this PR and follow these steps:
- add locations to [Vulnerabilities](https://github.com/aquasecurity/trivy/blob/8ec3938e01a93855503e3400eae9831abbb5de4a/rpc/common/service.proto#L110-L135) (use 26 number). Use [Location](https://github.com/DmitriyLewen/trivy/blob/84dc645df01ed5a9c7bdc141888a59a9b7f52fd2/rpc/common/service.proto#L64-L67) struct.
- run `mage fmt`.
- run `mage:protoc`.
- add `Locations` to `Vulnerabilties` [as for Packages](https://github.com/DmitriyLewen/trivy/blob/84dc645df01ed5a9c7bdc141888a59a9b7f52fd2/pkg/rpc/convert.go#L68) ( use `ConvertToRPCLocations` and `ConvertFromRPCLocation` functions)
- add Locations in testcase in `convert_test.go`.
- recheck that you don't need new test golden files in integration tests.

**Review by DmitriyLewen — APPROVED** [review]
@parvez0 thanks for your work!

@knqyf263 I approved PR. Take a looks, when you have time, please.

**Comment by knqyf263** [comment]
Originally, our policy was to include only minimal data in reports. It may be better to embed [Package](https://github.com/aquasecurity/trivy/blob/30cc88fa87ba69c64f09885c23f36663fa8aa5d1/pkg/fanal/types/artifact.go#L68-L105) in [DetectedVulnerability](https://github.com/aquasecurity/trivy/blob/30cc88fa87ba69c64f09885c23f36663fa8aa5d1/pkg/types/vulnerability.go#L9-L31) than to gradually add more and more attributes.

**Comment by DmitriyLewen** [comment]
sounds logical. 
@parvez0  Do you have the time and opportunity to do this?

**Comment by parvez0** [comment]
Hey @DmitriyLewen, I'll take a look at this and update

**Comment by knqyf263** [comment]
We're so sorry. We discussed this offline and found out that embedding packages will get lots of duplicate data if multiple vulnerabilities are found in one package. We need to carefully think about it.

**Comment by parvez0** [comment]
Hey @knqyf263, I had the same question and was just about to reach out to you. Please let me know if you've settled on an approach for this and need my assistance to finalize it. In the meantime, I'll start looking into other issues that I can address.

**Comment by knqyf263** [comment]
@parvez0 Thanks for your understanding. I implemented [my idea](https://github.com/aquasecurity/trivy/pull/6583) and would collect feedback from other maintainers.
